diff --git a/lib/vquic/ngtcp2.c b/lib/vquic/ngtcp2.c
index 9e7f99e87..6b827a32b 100644
--- a/lib/vquic/ngtcp2.c
+++ b/lib/vquic/ngtcp2.c
@@ -869,6 +869,7 @@ CURLcode Curl_quic_connect(struct Curl_easy *data,
   qs->conn_ref.get_conn = get_conn;
   qs->conn_ref.user_data = qs;
 
+  conn->ssl[sockindex].state = ssl_connection_negotiating;
   return CURLE_OK;
 }
 
@@ -961,8 +962,10 @@ void Curl_quic_disconnect(struct Curl_easy *data,
                           int tempindex)
 {
   (void)data;
-  if(conn->transport == TRNSPRT_QUIC)
+  if(conn->transport == TRNSPRT_QUIC) {
     qs_disconnect(&conn->hequic[tempindex]);
+    conn->ssl[tempindex].state = ssl_connection_none;
+  }
 }
 
 static CURLcode ng_disconnect(struct Curl_easy *data,
@@ -1728,8 +1731,10 @@ CURLcode Curl_quic_is_connected(struct Curl_easy *data,
 
   if(ngtcp2_conn_get_handshake_completed(qs->qconn)) {
     result = ng_has_connected(data, conn, sockindex);
-    if(!result)
+    if(!result) {
+      conn->ssl[sockindex].state = ssl_connection_complete;
       *done = TRUE;
+    }
   }
 
   return result;
