diff --git a/lib/vquic/ngtcp2.c b/lib/vquic/ngtcp2.c
index 6b827a32b..4b7e2eb8d 100644
--- a/lib/vquic/ngtcp2.c
+++ b/lib/vquic/ngtcp2.c
@@ -1017,6 +1017,9 @@ static int cb_h3_stream_close(nghttp3_conn *conn, int64_t stream_id,
 {
   struct Curl_easy *data = stream_user_data;
   struct HTTP *stream = data->req.p.http;
+  if (stream == NULL) {
+      return 0;
+  }
   (void)conn;
   (void)stream_id;
   (void)app_error_code;
@@ -1123,6 +1126,9 @@ static int cb_h3_end_headers(nghttp3_conn *conn, int64_t stream_id,
 {
   struct Curl_easy *data = stream_user_data;
   struct HTTP *stream = data->req.p.http;
+  if (stream == NULL) {
+      return 0;
+  }
   CURLcode result = CURLE_OK;
   (void)conn;
   (void)stream_id;
@@ -1152,6 +1158,9 @@ static int cb_h3_recv_header(nghttp3_conn *conn, int64_t stream_id,
   nghttp3_vec h3val = nghttp3_rcbuf_get_buf(value);
   struct Curl_easy *data = stream_user_data;
   struct HTTP *stream = data->req.p.http;
+  if (stream == NULL) {
+    return 0;
+  }
   CURLcode result = CURLE_OK;
   (void)conn;
   (void)stream_id;
@@ -1411,6 +1420,9 @@ static int cb_h3_acked_stream_data(nghttp3_conn *conn, int64_t stream_id,
 {
   struct Curl_easy *data = stream_user_data;
   struct HTTP *stream = data->req.p.http;
+  if (stream == NULL) {
+    return 0;
+  }
   (void)user_data;
 
   if(!data->set.postfields) {
@@ -1450,6 +1462,10 @@ static nghttp3_ssize cb_h3_readfunction(nghttp3_conn *conn, int64_t stream_id,
     return 1;
   }
 
+  if (stream == NULL) {
+    return 0;
+  }
+
   if(stream->upload_len && H3_SEND_SIZE <= stream->h3out->used) {
     return NGHTTP3_ERR_WOULDBLOCK;
   }
