diff --git a/quiche/Cargo.toml b/quiche/Cargo.toml
index f323b9f1..6ce1d10e 100644
--- a/quiche/Cargo.toml
+++ b/quiche/Cargo.toml
@@ -74,4 +74,5 @@ mio = { version = "0.8", features = ["net", "os-poll"] }
 url = "1"
 
 [lib]
-crate-type = ["lib", "staticlib", "cdylib"]
+#crate-type = ["lib", "staticlib", "cdylib"]
+crate-type = ["staticlib"]
diff --git a/quiche/src/build.rs b/quiche/src/build.rs
index d94bed5b..f8c69b0f 100644
--- a/quiche/src/build.rs
+++ b/quiche/src/build.rs
@@ -245,11 +245,11 @@ fn main() {
         println!("cargo:rustc-link-lib=static=crypto");
     }
 
-    // MacOS: Allow cdylib to link with undefined symbols
-    let target_os = std::env::var("CARGO_CFG_TARGET_OS").unwrap();
-    if target_os == "macos" {
-        println!("cargo:rustc-cdylib-link-arg=-Wl,-undefined,dynamic_lookup");
-    }
+    // // MacOS: Allow cdylib to link with undefined symbols
+    // let target_os = std::env::var("CARGO_CFG_TARGET_OS").unwrap();
+    // if target_os == "macos" {
+    //     println!("cargo:rustc-cdylib-link-arg=-Wl,-undefined,dynamic_lookup");
+    // }
 
     if cfg!(feature = "pkg-config-meta") {
         write_pkg_config();
