diff --git a/CMakeLists.txt b/CMakeLists.txt
index aaf3bf1..1a98244 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -530,6 +530,24 @@ elseif(MSVC)
     "$<$<COMPILE_LANGUAGE:C,CXX>:-W4;-WX;${MSVC_DISABLED_WARNINGS_LIST}>")
 endif()
 
+# Begin AdGuard compile options
+if(WIN32)
+  # Minimum supported Windows API is 6.3 (Windows 8.1)
+  add_definitions(-D_WIN32_WINNT=0x0603)
+  add_definitions(-D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE)
+  if(MSVC)
+    # Force to compile everything with /MT
+    set(CMAKE_MSVC_RUNTIME_LIBRARY "MultiThreaded")
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /Zi /EHs-c-")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zi /EHs-c-")
+  else()
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fno-strict-aliasing -fno-exceptions")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fno-strict-aliasing -fno-exceptions")
+    link_libraries(-static)
+  endif()
+endif()
+# End AdGuard compile options
+
 if(OPENSSL_ASM)
   set(CRYPTO_SOURCES_ASM_USED ${CRYPTO_SOURCES_ASM})
   set(BCM_SOURCES_ASM_USED ${BCM_SOURCES_ASM})
