diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1a98244..db6565d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -549,6 +549,9 @@ endif()
 # End AdGuard compile options
 
 if(OPENSSL_ASM)
+  if(NOT (ANDROID OR CMAKE_SIZEOF_VOID_P EQUAL 8))
+    list(REMOVE_ITEM CRYPTO_SOURCES_ASM crypto/curve25519/asm/x25519-asm-arm.S)
+  endif()
   set(CRYPTO_SOURCES_ASM_USED ${CRYPTO_SOURCES_ASM})
   set(BCM_SOURCES_ASM_USED ${BCM_SOURCES_ASM})
   set(TEST_SUPPORT_SOURCES_ASM_USED ${TEST_SUPPORT_SOURCES_ASM})
diff --git a/crypto/curve25519/internal.h b/crypto/curve25519/internal.h
index c3bc6fe..a28a774 100644
--- a/crypto/curve25519/internal.h
+++ b/crypto/curve25519/internal.h
@@ -21,7 +21,7 @@
 
 BSSL_NAMESPACE_BEGIN
 
-#if defined(OPENSSL_ARM) && !defined(OPENSSL_NO_ASM) && !defined(OPENSSL_APPLE)
+#if defined(OPENSSL_ARM) && !defined(OPENSSL_NO_ASM) && !defined(OPENSSL_APPLE) && defined(__clang__)
 #define BORINGSSL_X25519_NEON
 
 // x25519_NEON is defined in asm/x25519-arm.S.
