diff --git a/CMakeLists.txt b/CMakeLists.txt
index db6565d..9ce6c63 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -294,6 +294,10 @@ if(OPENSSL_NO_SSE2_FOR_TESTING)
   add_definitions(-DOPENSSL_NO_SSE2_FOR_TESTING)
 endif()
 
+if(WIN32 AND OPENSSL_NO_ASM)
+  add_definitions(-DOPENSSL_STATIC_ARMCAP)
+endif()
+
 # Build a custom libc++. This must be the first target, so that it is linked to
 # every subsequent target.
 if(USE_CUSTOM_LIBCXX)
@@ -525,6 +529,8 @@ elseif(MSVC)
       "-wd4702" # unreachable code; MSVC's warning is too aggressive. See
                 # https://crbug.com/385161043
       "-wd4706" # assignment within conditional expression
+      "-wd4132" # const object should be initialized
+      "-wd4146" # unary minus operator applied to unsigned type, result still unsigned
       )
   add_compile_options(
     "$<$<COMPILE_LANGUAGE:C,CXX>:-W4;-WX;${MSVC_DISABLED_WARNINGS_LIST}>")
