diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6d71eda..a8ca0e1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -902,7 +902,11 @@ endif()
 
 if(INSTALL_ENABLED)
   install(TARGETS crypto ssl EXPORT OpenSSLTargets)
-  install(TARGETS bssl)
+  if (APPLE)
+    install(TARGETS bssl BUNDLE DESTINATION .)
+  else ()
+    install(TARGETS bssl)
+  endif()
   install(DIRECTORY include/ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
   install(EXPORT OpenSSLTargets
           FILE OpenSSLTargets.cmake
