diff --git a/ssl/extensions.cc b/ssl/extensions.cc
index 5ee2802..30c2d81 100644
--- a/ssl/extensions.cc
+++ b/ssl/extensions.cc
@@ -3294,10 +3294,12 @@ bool ssl_setup_extension_permutation(SSL_HANDSHAKE *hs) {
   for (size_t i = 0; i < kNumExtensions; i++) {
     permutation[i] = i;
   }
-  for (size_t i = kNumExtensions - 1; i > 0; i--) {
-    // Set element |i| to a randomly-selected element 0 <= j <= i.
-    std::swap(permutation[i], permutation[seeds[i - 1] % (i + 1)]);
+  for (size_t i = kNumExtensions - 1; i > 1; i--) {
+    // Set element |i| to a randomly-selected element 1 <= j <= i.
+    std::swap(permutation[i], permutation[(seeds[i - 1] % i) + 1]);
   }
+  // SNI always last
+  std::swap(permutation[0], permutation[kNumExtensions - 1]);
   hs->extension_permutation = std::move(permutation);
   return true;
 }
