diff --git a/ssl/extensions.cc b/ssl/extensions.cc
index 5ee2802..c58ba4f 100644
--- a/ssl/extensions.cc
+++ b/ssl/extensions.cc
@@ -3298,6 +3298,19 @@ bool ssl_setup_extension_permutation(SSL_HANDSHAKE *hs) {
     // Set element |i| to a randomly-selected element 0 <= j <= i.
     std::swap(permutation[i], permutation[seeds[i - 1] % (i + 1)]);
   }
+  size_t server_name_idx = 0, server_name_pos = 0;
+  size_t key_share_idx = 0, key_share_pos = 0;
+  for (size_t i = 0; i < kNumExtensions; i++) {
+    if (kExtensions[i].value == TLSEXT_TYPE_key_share) key_share_idx = i;
+    if (kExtensions[i].value == TLSEXT_TYPE_server_name) server_name_idx = i;
+  }
+  for (size_t i = 0; i < kNumExtensions; i++) {
+    if (permutation[i] == key_share_idx) key_share_pos = i;
+    if (permutation[i] == server_name_idx) server_name_pos = i;
+  }
+  if (key_share_pos > server_name_pos) {
+    std::swap(permutation[key_share_pos], permutation[server_name_pos]);
+  }
   hs->extension_permutation = std::move(permutation);
   return true;
 }
