diff --git a/crypto/x509v3/v3_utl.c b/crypto/x509v3/v3_utl.c
index bbc82e283..243e8a954 100644
--- a/crypto/x509v3/v3_utl.c
+++ b/crypto/x509v3/v3_utl.c
@@ -790,7 +790,7 @@ static int wildcard_match(const unsigned char *prefix, size_t prefix_len,
   // Check that the part matched by the wildcard contains only
   // permitted characters and only matches a single label.
   for (p = wildcard_start; p != wildcard_end; ++p) {
-    if (!OPENSSL_isalnum(*p) && *p != '-') {
+    if (!OPENSSL_isalnum(*p) && *p != '-' && *p != '_') {
       return 0;
     }
   }
@@ -1011,7 +1011,9 @@ static int do_x509_check(X509 *x, const char *chk, size_t chklen,
       }
     }
     GENERAL_NAMES_free(gens);
-    return rv;
+    // Added by AdGuard. Google removed X509_CHECK_FLAG_ALWAYS_CHECK_SUBJECT.
+    if (rv != 0)
+      return rv;
   }
 
   // We're done if CN-ID is not pertinent
