diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0bf394d77..dc7d5ae38 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -209,6 +209,9 @@ elseif(MSVC)
               # possible loss of data
       "C4267" # conversion from 'size_t' to 'int', possible loss of data
       "C4706" # assignment within conditional expression
+      "C5105" # AdGuard: UB warning treated as error in WinBase.h on VS v17.3.6
+      "C5105" # AdGuard: UB warning treated as error in WinBase.h on VS v17.3.6
+      "C4746" # AdGuard: arm64 build fails with warning treated as error: "volatile access of 'kBoringSSLBinaryTag' is subject to /volatile:<iso|ms> setting; consider using __iso_volatile_load/store intrinsic functions"
       )
   string(REPLACE "C" " -wd" MSVC_DISABLED_WARNINGS_STR
                             ${MSVC_DISABLED_WARNINGS_LIST})
