diff --git a/CMake/FindNGHTTP2.cmake b/CMake/FindNGHTTP2.cmake
index 6d70c4a04..81d0aa576 100644
--- a/CMake/FindNGHTTP2.cmake
+++ b/CMake/FindNGHTTP2.cmake
@@ -25,7 +25,7 @@ include(FindPackageHandleStandardArgs)
 
 find_path(NGHTTP2_INCLUDE_DIR "nghttp2/nghttp2.h")
 
-find_library(NGHTTP2_LIBRARY NAMES nghttp2)
+find_library(NGHTTP2_LIBRARY NAMES nghttp2 nghttp2_static)
 
 find_package_handle_standard_args(NGHTTP2
     FOUND_VAR
diff --git a/CMake/FindNGHTTP3.cmake b/CMake/FindNGHTTP3.cmake
index 8d8ebc1b1..fe81748b3 100644
--- a/CMake/FindNGHTTP3.cmake
+++ b/CMake/FindNGHTTP3.cmake
@@ -52,7 +52,7 @@ find_path(NGHTTP3_INCLUDE_DIR nghttp3/nghttp3.h
     ${PC_NGHTTP3_INCLUDE_DIRS}
 )
 
-find_library(NGHTTP3_LIBRARY NAMES nghttp3
+find_library(NGHTTP3_LIBRARY NAMES nghttp3 nghttp3_static
   HINTS
     ${PC_NGHTTP3_LIBDIR}
     ${PC_NGHTTP3_LIBRARY_DIRS}
diff --git a/CMake/FindNGTCP2.cmake b/CMake/FindNGTCP2.cmake
index 61e54c2d6..5b78af8ae 100644
--- a/CMake/FindNGTCP2.cmake
+++ b/CMake/FindNGTCP2.cmake
@@ -58,7 +58,7 @@ find_path(NGTCP2_INCLUDE_DIR ngtcp2/ngtcp2.h
     ${PC_NGTCP2_INCLUDE_DIRS}
 )
 
-find_library(NGTCP2_LIBRARY NAMES ngtcp2
+find_library(NGTCP2_LIBRARY NAMES ngtcp2 ngtcp2_static
   HINTS
     ${PC_NGTCP2_LIBDIR}
     ${PC_NGTCP2_LIBRARY_DIRS}
@@ -87,6 +87,7 @@ if(NGTCP2_FIND_COMPONENTS)
     find_library(${_crypto_library}_LIBRARY
       NAMES
         ${_crypto_library}
+        ${_crypto_library}_static
       HINTS
         ${PC_${_crypto_library}_LIBDIR}
         ${PC_${_crypto_library}_LIBRARY_DIRS}
