diff --git a/lib/select.c b/lib/select.c
index 3b8d468bd..6edc5fb12 100644
--- a/lib/select.c
+++ b/lib/select.c
@@ -373,8 +373,10 @@ int Curl_poll(struct pollfd ufds[], unsigned int nfds, timediff_t timeout_ms)
     if(FD_ISSET(ufds[i].fd, &fds_read)) {
       if(ufds[i].events & POLLRDNORM)
         ufds[i].revents |= POLLRDNORM;
+#ifndef _WIN32
       if(ufds[i].events & POLLIN)
         ufds[i].revents |= POLLIN;
+#endif
     }
     if(FD_ISSET(ufds[i].fd, &fds_write)) {
       if(ufds[i].events & POLLWRNORM)
