diff --git a/signal.c b/signal.c
index 89f5fc17..340d91d1 100644
--- a/signal.c
+++ b/signal.c
@@ -187,6 +187,11 @@ evsig_cb(evutil_socket_t fd, short what, void *arg)
 int
 evsig_init_(struct event_base *base)
 {
+#ifdef _WIN32
+	// Not used by AdGuard
+	return 0;
+#endif
+
 	/*
 	 * Our signal handler is going to write to one end of the socket
 	 * pair to wake up our event loop.  The event loop then scans for
