diff --git a/bufferevent_sock.c b/bufferevent_sock.c
index 4c3c4b7b..4f1168e1 100644
--- a/bufferevent_sock.c
+++ b/bufferevent_sock.c
@@ -451,6 +451,10 @@ bufferevent_socket_connect(struct bufferevent *bev,
 	goto done;

 freesock:
+	result = evutil_socket_geterror(fd);
+	if (result == 0) {
+		result = -1;
+	}
 	if (ownfd)
 		evutil_closesocket(fd);
 done:
diff --git a/include/event2/bufferevent.h b/include/event2/bufferevent.h
index 74479be2..8f78d95d 100644
--- a/include/event2/bufferevent.h
+++ b/include/event2/bufferevent.h
@@ -215,7 +215,8 @@ struct bufferevent *bufferevent_socket_new(struct event_base *base, evutil_socke
        bufferevent_socket_new().
    @param addr the address we should connect to
    @param socklen The length of the address
-   @return 0 on success, -1 on failure.
+   @return 0 on success, system error code on failure
+         (the describing message may be retrieved via `evutil_socket_error_to_string`).
  */
 EVENT2_EXPORT_SYMBOL
 int bufferevent_socket_connect(struct bufferevent *, const struct sockaddr *, int);
